File: easy/array/single_number.java
import java.util.Arrays;

/*
https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/549/
Given a non-empty array of integers nums, every element appears twice except for one.
Find that single one.
*/

class Solution{
    public int singleNumber(int[] nums){
        int xor_value = 0;

        for(int i=0; i<nums.length; i++){
            xor_value ^= nums[i];
        }

        return xor_value;
    }
}

/* Solution using sorted array
class Solution{
    public int singleNumber(int[] nums){
        Arrays.sort(nums);
        int i;

        for(i=0; i<nums.length-2; i+=2){
            if(nums[i] != nums[i+1]) return nums[i];
        }

        return nums[i];
    }
}
*/

File: easy/binary/single_number.js
/* https://leetcode.com/problems/single-number/
Given a non-empty array of integers nums,
every element appears twice except for one. Find that single one.
You must implement a solution with a linear runtime complexity and use only constant extra space.
*/

/**
 * @param {number[]} nums
 * @return {number}
 */
const singleNumber = function (nums) {
  return nums.reduce((prev, curr) => prev ^ curr);
};

/* Hash map solution, O(n) time and O(n) space
const singleNumber = function (nums) {
  const hash = {};

  nums.every((num) => {
    hash[num] ? delete hash[num] : hash[num] = 1;
  });

  return Object.keys(hash)[0];
};
*/

/* Initial Solution, O(2n) time and O(1) space
const singleNumber = function (nums) {
  nums = nums.sort();

  for (let i = 0; i < nums.length - 1; i += 2) {
    if (nums[i + 1] !== nums[i]) return nums[i];
  }

  return nums[nums.length - 1];
};
*/
